src/ngx_http_lua_shdict.c

```diff
@@ -1823,9 +1957,35 @@ ngx_http_lua_ffi_shdict_incr(ngx_shm_zone_t *zone, u_char *key,
     dd("shdict lookup returned %d", (int) rc);
 
     if (rc == NGX_DECLINED || rc == NGX_DONE) {
-        ngx_shmtx_unlock(&ctx->shpool->mutex);
-        *err = "not found";
-        return NGX_ERROR;
+        if (arg_num == 0) {
+            ngx_shmtx_unlock(&ctx->shpool->mutex);
+            *err = "not found";
+            return NGX_ERROR;
+        }
+
+        /* add value */
+        num = *value + init;
+
+        if (rc == NGX_DONE) {
+
+            if ((size_t) sd->value_len == sizeof(double)) {
+                ngx_log_debug0(NGX_LOG_DEBUG_HTTP, ctx->log, 0,
+                               "lua shared dict incr: found old entry and value "
+                               "size matched, reusing it");
+
+                ngx_queue_remove(&sd->queue);
+                ngx_queue_insert_head(&ctx->sh->queue, &sd->queue);
+
+                dd("go to setvalue");
+                goto setvalue;
+            }
+
+            dd("go to remove");
+            goto remove;
```

Comment:
I think if the key already takes a non-numerical value, then we should really return an error instead of initializing with `init`. I think `init` should only be used when the key does not exist.